src/librbd/api/Trash.cc

```diff
@@ -116,27 +118,18 @@ int enable_mirroring(IoCtx &io_ctx, const std::string &image_id) {
 } // anonymous namespace
 
 template <typename I>
-int Trash<I>::move(librados::IoCtx &io_ctx, rbd_trash_image_source_t source,
-                   const std::string &image_name, uint64_t delay) {
+int Trash<I>::move_by_id(librados::IoCtx &io_ctx, rbd_trash_image_source_t source,
+                   const std::string &image_id, const std::string &image_name,
+                   uint64_t delay, bool check_for_watchers) {
   CephContext *cct((CephContext *)io_ctx.cct());
-  ldout(cct, 20) << "trash_move " << &io_ctx << " " << image_name
+  ldout(cct, 20) << &io_ctx << " " << (image_id.empty() ? image_name : image_id)
                  << dendl;
 
-  // try to get image id from the directory
-  std::string image_id;
-  int r = cls_client::dir_get_id(&io_ctx, RBD_DIRECTORY, image_name,
-                                 &image_id);
-  if (r < 0 && r != -ENOENT) {
-    lderr(cct) << "failed to retrieve image id: " << cpp_strerror(r) << dendl;
-    return r;
-  }
-
   ImageCtx *ictx = new ImageCtx((image_id.empty() ? image_name : ""),
                                 image_id, nullptr, io_ctx, false);
-  r = ictx->state->open(OPEN_FLAG_SKIP_OPEN_PARENT);
-  if (r == -ENOENT) {
-    return r;
-  } else if (r < 0) {
+  int r = ictx->state->open(OPEN_FLAG_SKIP_OPEN_PARENT);
+
+  if (r < 0 && r != -ENOENT) {
     lderr(cct) << "failed to open image: " << cpp_strerror(r) << dendl;
```

Comment:
need to delete `ictx` before return in this case

Reply:
The synchronous version of ```ImageState::open``` will automatically delete the associated ```ImageCtx``` if it fails.

Reply:
Ah, forgot about this.